# https://metanit.com/python/tutorial/2.1.php
# Основы Python

### Введение в написание программ
Программа на языке Python состоит из набора инструкций. Каждая инструкция помещается на новую строку. Например:

In [1]:
print(2 + 3) 
print("Hello")

5
Hello


Большую роль в Python играют отступы. Неправильно поставленный отступ фактически является ошибкой. Например, в следующем случае мы получим ошибку, хотя код будет практически аналогичен приведенному выше:

In [2]:
print(2 + 3) 
    print("Hello")

IndentationError: unexpected indent (2234453416.py, line 2)

Поэтому стоит помещать новые инструкции сначала строки. В этом одно из важных отличий пайтона от других языков программирования, как C# или Java.

Однако стоит учитывать, что некоторые конструкции языка могут состоять из нескольких строк. Например, условная конструкция if:

In [3]:
if 1 < 2:
    print("Hello")

Hello


В данном случае если 1 меньше 2, то выводится строка "Hello". И здесь уже должен быть отступ, так как инструкция print("Hello") используется не сама по себе, а как часть условной конструкции if. Причем отступ, согласно руководству по оформлению кода, желательно делать из такого количество пробелов, которое кратно 4 (то есть 4, 8, 16 и т.д.) Хотя если отступов будет не 4, а 5, то программа также будет работать.

Таких конструкций не так много, поэтому особой путаницы по поводу где надо, а где не надо ставить пробелы, не должно возникнуть.

### Регистрозависимость

Python - регистрозависимый язык, поэтому выражения print и Print или PRINT представляют разные выражения. И если вместо метода print для вывода на консоль мы попробуем использовать метод Print:

In [4]:
Print("Hello World")

NameError: name 'Print' is not defined

то у нас ничего не получится.

Для отметки, что делает тот или иной участок кода, применяются комментарии. При трансляции и выполнении программы интерпретатор игнорирует комментарии, поэтому они не оказывают никакого влияния на работу программы. Комментарии в Python бывают блочные и строчные.

Строчные коментарии предваряются знаком решетки `- #`. Они могут располагаться на отдельной строке:

In [5]:
# Вывод на консоль 
# сообщения Hello World
print("Hello World")

Hello World


Любой набор символов после знака `#` представляет комментарий. То есть в примее выше первые две строки кода являются комментариями.

Также они могут располагаться на той же строке, что и инструкции языка, после выполняемых инструкций:

In [6]:
print("Hello World")  # Вывод сообщения на консоль

Hello World


В блочных коментариях до и после текста комментария ставятся три одинарные кавычки: '''текст комментария'''. Например:

In [7]:
''' 
    Вывод на консоль
    сообщения Hello World
'''
print("Hello World")

Hello World


### Основные функции

Python предоставляет ряд встроенных функций. Некоторые из них используются очень часто, особенно на начальных этапах изучения языка, поэтому рассмотрим их.

Основной функцией для вывода информации на консоль является функция print(). В качестве аргумента в эту функцию передается строка, которую мы хотим вывести:

In [8]:
print("Hello Python")

Hello Python


Если же нам необходимо вывести несколько значений на консоль, то мы можем передать их в функцию print через запятую:

In [9]:
print("Full name:", "Tom", "Smith")

Full name: Tom Smith


В итоге все переданные значения склеятся через пробелы в одну строку:

Если функция print отвечает за вывод, то функция input отвечает за ввод информации. В качестве необязательного параметра эта функция принимает приглашение к вводу и возвращает введенную строку, которую мы можем сохранить в переменную:

In [10]:
name = input("Введите имя: ")
print("Привет", name)

Введите имя: yu
Привет yu


Консольный вывод:
```
Введите имя: Евгений
Привет Евгений
```

# Переменные и типы данных

### Переменные

Переменные предназначены для хранения данных. Название переменной в Python должно начинаться с алфавитного символа или со знака подчеркивания и может содержать алфавитно-цифровые символы и знак подчеркивания. И кроме того, название переменной не должно совпадать с названием ключевых слов языка Python. Ключевых слов не так много, их легко запомнить:
```
False      await      else       import     pass
None       break      except     in         raise
True       class      finally    is         return
and        continue   for        lambda     try
as         def        from       nonlocal   while
assert     del        global     not        with
async      elif       if         or         yield
```
Например, создадим переменную:

In [11]:
name = "Tom"

Здесь определена переменная name, которая хранит строку "Tom".

В пайтоне применяется два типа наименования переменных: `camel case и underscore notation`

`Camel case` подразумевает, что каждое новое подслово в наименовании переменной начинается с большой буквы. Например:

In [12]:
userName = "Tom"

`Underscore notation` подразумевает, что подслова в наименовании переменной разделяются знаком подчеркивания. Например:

In [13]:
user_name = "Tom"

И также надо учитывать регистрозависимость, поэтому переменные name и Name будут представлять разные объекты.

In [14]:
# две разные переменные
name = "Tom"
Name = "Tom"

Определив переменную, мы можем использовать в программе. Например, попытаться вывести ее содержимое на консоль с помощью встроенной функции print:

In [15]:
name = "Tom"  # определение переменной name
print(name)   # вывод значения переменной name на консоль

Tom


Например, определение и применение переменной в среде PyCharm:

Отличительной особенностью переменной является то, что мы можем менять ее значение в течение работы программы:

In [16]:
name = "Tom"  # переменной name равна "Tom"
print(name)   # выводит: Tom
name = "Bob"  # меняем значение на "Bob"
print(name)   # выводит: Bob

Tom
Bob


### Типы данных

Переменная хранит данные одного из типов данных. В Python существует множество различных типов данных. В данном случае рассмотрим только самые базовые типы: 

- bool, 
- int, 
- float, 
- complex,
- str
- ...

### Логические значения

Тип `bool` представляет два логических значения: __True__ (верно, истина) или __False__ (неверно, ложь). Значение True служит для того, чтобы показать, что что-то истинно. Тогда как значение False, наоборот, показывает, что что-то ложно. Пример переменных данного типа:

In [17]:
isMarried = False
print(isMarried)    # False
 
isAlive = True
print(isAlive)      # True

False
True


### Целые числа

Тип int представляет целое число, например, 1, 4, 8, 50. Пример

In [18]:
age = 21
print("Возраст:", age)    # Возраст: 21
 
count = 15
print("Количество:", count) # Количество: 15

Возраст: 21
Количество: 15


По умолчанию стандартные числа расцениваются как числа в десятичной системе. Но Python также поддерживает числа в двоичной, восьмеричной и шестнадцатеричной системах.

Для указания, что число представляет двоичную систему, перед числом ставится префикс `0b`:

In [19]:
a = 0b11
b = 0b1011
c = 0b100001
print(a)    # 3 в десятичной системе
print(b)    # 11 в десятичной системе
print(c)    # 33 в десятичной системе

3
11
33


Для указания, что число представляет восьмеричную систему, перед числом ставится префикс `0o`:

In [20]:
a = 0o7
b = 0o11
c = 0o17
print(a)    # 7 в десятичной системе
print(b)    # 9 в десятичной системе
print(c)    # 15 в десятичной системе

7
9
15


Для указания, что число представляет шестнадцатеричную систему, перед числом ставится префикс `0x`:

In [21]:
a = 0x0A
b = 0xFF
c = 0xA1
print(a)    # 10 в десятичной системе
print(b)    # 255 в десятичной системе
print(c)    # 161 в десятичной системе

10
255
161


Стоит отметить, что в какой-бы системе мы не передали число в функцию print для вывода на консоль, оно по умолчанию будет выводиться в десятичной системе.
Дробные числа

Тип float представляет число с плавающей точкой, например, 1.2 или 34.76. В качесте разделителя целой и дробной частей используется точка.

In [22]:
height = 1.68
pi = 3.14
weight = 68.
print(height)   # 1.68
print(pi)       # 3.14
print(weight)   # 68.0

1.68
3.14
68.0


Число с плавающей точкой можно определять в экспоненциальной записи:

In [23]:
x = 3.9e3
print(x)  # 3900.0
 
x = 3.9e-3
print(x)  # 0.0039

3900.0
0.0039


Число float может иметь только 18 значимых символов. Так, в данном случае используются только два символа - 3.9. И если число слишком велико или слишком мало, то мы можем записывать число в подобной нотации, используя экспоненту. Число после экспоненты указывает степень числа 10, на которое надо умножить основное число - 3.9.
Комплексные числа

Тип `complex` представляет комплексные числа в формате вещественная_часть+мнимая_частьj - после мнимой части указывается суффикс j

In [24]:
complexNumber = 1+2j
print(complexNumber)   # (1+2j)

(1+2j)


### Строки

Тип str представляет строки. Строка представляет последовательность символов, заключенную в одинарные или двойные кавычки, например "hello" и 'hello'. В Python 3.x строки представляют набор символов в кодировке Unicode

In [25]:
message = "Hello World!"
print(message)  # Hello World!
 
name = 'Tom'
print(name)  # Tom

Hello World!
Tom


При этом, если строка имеет много символов, ее можно разбить на части и эти части разместить на разных строках кода. В этом случае вся строка заключается в круглые скобки, а ее отдельные части - в кавычки:

In [26]:
text = ("Laudate omnes gentes laudate "
        "Magnificat in secula ")
print(text)

Laudate omnes gentes laudate Magnificat in secula 


Если же мы хотим определить многострочный текст, то такой текст заключается в тройные двойные или одинарные кавычки:

In [27]:
'''
Это комментарий
'''
text = '''Laudate omnes gentes laudate
Magnificat in secula
Et anima mea laudate
Magnificat in secula 
'''
print(text)

Laudate omnes gentes laudate
Magnificat in secula
Et anima mea laudate
Magnificat in secula 



При использовани тройных одинарных кавычек не стоит путать их с комментариями: если текст в тройных одинарных кавычках присваивается переменной, то это строка, а не комментарий.
Управляющие последовательности в строке

Строка может содержать ряд специальных символов - управляющих последовательностей. Некоторые из них:

-    `\\`: позволяет добавить внутрь строки слеш

-    `\'`: позволяет добавить внутрь строки одинарную кавычку

-    `\"`: позволяет добавить внутрь строки двойную кавычку

-    `\n` осуществляет переход на новую строку

-    `\t`: добавляет табуляцию (4 отступа)

Применим несколько последовательностей:

In [28]:
text = "Message:\n\"Hello World\""
print(text)

Message:
"Hello World"


Консольный вывод программы:
```
Message:
"Hello World"
```
Хотя подобные последовательности могут нам помочь в некоторых делах, например, поместить в строку кавычку, сделать табуляцию, перенос на другую строку. Но они также могут и мешать. Например:

In [29]:
path = "C:\python\name.txt"
print(path)

C:\python
ame.txt


Чтобы избежать подобной ситуации, перед строкой ставится символ `r`

```
C:\python
ame.txt
```
Чтобы избежать подобной ситуации, перед строкой ставится символ r

In [30]:
path = r"C:\python\name.txt"
print(path)

C:\python\name.txt


### Вставка значений в строку

Python позволяет встравивать в строку значения других переменных. Для этого внутри строки переменные размещаются в фигурных скобках {}, а перед всей строкой ставится символ f:

In [31]:
userName = "Tom"
userAge = 37
user = f"name: {userName}  age: {userAge}"
print(user)   # name: Tom  age: 37

name: Tom  age: 37


В данном случае на место __{userName}__ будет вставляться значение переменной userName. Аналогично на вместо {userAge} будет вставляться значение переменной userAge.
Динамическая типизация

Python является языком с динамической типизацией. А это значит, что переменная не привязана жестко к определенному типу.

Тип переменной определяется исходя из значения, которое ей присвоено. Так, при присвоении строки в двойных или одинарных кавычках переменная имеет тип str. При присвоении целого числа Python автоматически определяет тип переменной как int. Чтобы определить переменную как объект float, ей присваивается дробное число, в котором разделителем целой и дробной части является точка.

При этом в процессе работы программы мы можем изменить тип переменной, присвоив ей значение другого типа:

In [32]:
userId = "abc"  # тип str
print(userId)
 
userId = 234  # тип int
print(userId)

abc
234


С помощью встроенной функции __type()__ динамически можно узнать текущий тип переменной:

In [33]:
userId = "abc"      # тип str
print(type(userId)) # <class 'str'>
 
userId = 234        # тип int
print(type(userId)) # <class 'int'>

<class 'str'>
<class 'int'>


# Консольный ввод и вывод
### Вывод на консоль

Для вывода информации на консоль предназначена встроенная функция print(). При вызове этой функции ей в скобках передается выводимое значение:

In [34]:
print("Hello METANIT.COM")

Hello METANIT.COM


Данный код выведет нам на консоль строку "Hello METANIT.COM".

Отличительной особенностью этой функции является то, что по умолчанию она выводит значение на отдельной строке. Например:

In [35]:
print("Hello World")
print("Hello METANIT.COM")
print("Hello Python")

Hello World
Hello METANIT.COM
Hello Python


Здесь три вызова функции print() выводят некоторое сообщение. Причем при выводе на консоль каждое сообщение будет размещаться на отдельной строке:
```
Hello World
Hello METANIT.COM
Hello Python
```
Такое поведение не всегда удобно. Например, мы хотим, чтобы все значения выводились на одной строке. Для этого нам надо настроить поведение функции с помощью параметра end. Этот параметр задает символы, которые добавляются в конце к выводимой строке и . При применении параметра end вызов функции print() выглядит следующим образом:
```
print(значение, end = конечные_символы)
```
По умолчанию end равен символу "\n", который задает перевод на следующую строку. Собственно поэтому функция print по умолчанию выводит передаваемое ей значение на отдельной строке.

Теперь определим, чтобы функция не делала перевод на следующую строку, а выводила значение на той же строке:

In [36]:
print("Hello World", end=" ")
print("Hello METANIT.COM", end=" ")
print("Hello Python")

Hello World Hello METANIT.COM Hello Python


То есть теперь выводимые значения будут разделяться пробелом:
```
Hello World Hello METANIT.COM Hello Python
```
Причем это может быть не один символ, а набор символов:

In [37]:
print("Hello World", end=" and ")
print("Hello METANIT.COM", end=" and ")
print("Hello Python")

Hello World and Hello METANIT.COM and Hello Python


В данном случае выводимые сообщения будут отделяться символами " and ":
```
Hello World and Hello METANIT.COM and Hello Python
```
### Консольный ввод

Наряду с выводом на консоль мы можем получать ввод пользователя с консоли, получать вводимые данные. Для этого в Python определена функция input(). В эту функцию передается приглашение к вводу. А результат ввода мы можем сохранить в переменную. Например, определим код для ввода пользователем имени:

In [38]:
name = input("Введите свое имя: ")
print(f"Ваше имя: {name}")

Введите свое имя: rty
Ваше имя: rty


В данном случае в функцию input() передается приглашение к вводу в виде строки "Введите свое имя: ". Результат функции - результат ввода пользователя передается в переменную name. Затем мы можем вывести значение этой переменной на консоль с помощью функции print(). Пример работы кода:
```
Введите свое имя: Eugene
Ваше имя: Eugene
```
Еще пример с вводом нескольких значений:

In [39]:
name = input("Your name: ")
age = input("Your age: ")
print(f"Name: {name}  Age: {age}")

Your name: tre
Your age: 12
Name: tre  Age: 12


Пример работы программы:
```
Your name: Tom
Your age: 37
Name: Tom  Age: 37
```
Стоит учитывать, что все введенные значения рассматриваются как значения типа str, то есть строки. И даже если мы вводим число, как в втором случае в коде выше, то Python все равно будет рассматривать введенное значение как строку, а не как число.

# Арифметические операции с числами

Python поддерживает все распространенные арифметические операции:

- `+` Сложение двух чисел:

In [40]:
print(6 + 2)  # 8

8


- `-` Вычитание двух чисел:

In [41]:
print(6 - 2)  # 4

4


- `*` Умножение двух чисел:

In [42]:
    print(6 * 2)  # 12

12


- `/` Деление двух чисел:

In [43]:
    print(6 / 2)  # 3.0

3.0


- `//` Целочисленное деление двух чисел:

In [44]:
print(7 / 2)  # 3.5
print(7 // 2)  # 3

3.5
3


Данная операция возвращает целочисленный результат деления, отбрасывая дробную часть

- `**` Возведение в степень:

In [45]:
print(6 ** 2)  # Возводим число 6 в степень 2. Результат - 36

36


- `%` Получение остатка от деления:

In [46]:
print(7 % 2)  # Получение остатка от деления числа 7 на 2. Результат - 1

1


В данном случае ближайшее число к 7, которое делится на 2 без остатка, это 6. Поэтому остаток от деления равен 7 - 6 = 1

При последовательном использовании нескольких арифметических операций их выполнение производится в соответствии с их приоритетом. В начале выполняются операции с большим приоритетом. Приоритеты операций в порядке убывания приведены в следующей таблице.

### Операции

- Направление `**`
- Справо налево `* / // %`
- Слева направо `+ -`


Пусть у нас выполняется следующее выражение:

In [48]:
number = 3 + 4 * 5 ** 2 + 7
print(number)  # 110

110


Здесь начале выполняется возведение в степень (5 ** 2) как операция с большим приоритетом, далее результат умножается на 4 (25 * 4), затем происходит сложение (3 + 100) и далее опять идет сложение (103 + 7).

Чтобы переопределить порядок операций, можно использовать скобки:

In [49]:
number = (3 + 4) * (5 ** 2 + 7)
print(number)  # 224

224


Следует отметить, что в арифметических операциях могут принимать участие как целые, так и дробные числа. Если в одной операции участвует целое число (int) и число с плавающей точкой (float), то целое число приводится к типу float.

### Арифметические операции с присвоением

Ряд специальных операций позволяют использовать присвоить результат операции первому операнду:

- `+=` Присвоение результата сложения

- `-=` Присвоение результата вычитания

- `*=` Присвоение результата умножения

- `/=` Присвоение результата от деления

- `//=` Присвоение результата целочисленного деления

- `**=` Присвоение степени числа

- `%=` Присвоение остатка от деления

Примеры операций:

In [50]:
number = 10
number += 5
print(number)  # 15
 
number -= 3
print(number)  # 12
 
number *= 4
print(number)  # 48

15
12
48


### Округление и функция round

При операциях с числами типа float надо учитывать, что результат операций с ними может быть не совсем точным. Например:

In [51]:
first_number = 2.0001
second_number = 5
third_number = first_number / second_number
print(third_number) # 0.40002000000000004

0.40002000000000004


В данном случае мы ожидаем получить число 0.40002, однако в конце через ряд нулей появляется еще какая-то четверка. Или еще одно выражение:

In [52]:
print(2.0001 + 0.1)  # 2.1001000000000003

2.1001000000000003


В случае выше для округления результата мы можем использовать встроенную функцию __round()__:

In [53]:
first_number = 2.0001
second_number = 0.1
third_number = first_number + second_number
print(round(third_number))  # 2

2


В функцию __round()__ передается число, которое надо округлить. Если в функцию передается одно число, как в примере выше, то оно округляется до целого.

Функция __round()__ также может принимать второе число, которое указывает, сколько знаков после запятой должно содержать получаемое число:

In [54]:
first_number = 2.0001
second_number = 0.1
third_number = first_number + second_number
print(round(third_number, 4))  # 2.1001

2.1001


В данном случае число third_number округляется до 4 знаков после запятой.

Если в функцию передается только одно значение - только округляемое число, оно округляется то ближайшего целого

Примеры округлений:

In [55]:
# округление до целого числа
print(round(2.49))  # 2 - округление до ближайшего целого 2
print(round(2.51))  # 3

2
3


Однако если округляемая часть равна одинаково удалена от двух целых чисел, то округление идет к ближайшему четному:

In [56]:
print(round(2.5))   # 2 - ближайшее четное
print(round(3.5))   # 4 - ближайшее четное

2
4


Округление производится до ближайшего кратного 10 в степени минус округляемая часть:

In [57]:
# округление до двух знаков после запятой
print(round(2.554, 2))      # 2.55
print(round(2.5551, 2))      # 2.56
print(round(2.554999, 2))   # 2.55
print(round(2.499, 2))      # 2.5

2.55
2.56
2.55
2.5


Однако следует учитывать, что функция round() не идеальный инструмент. Например, выше при округление до целых чисел применяется правило, согласно которому, если округляемая часть одинаково удалена от двух значений, то округление производится до ближайшего четного значения. В Python в связи с тем, что десятичная часть числа не может быть точно представлена в виде числа float, то это может приводить к некоторым не совсем ожидаемым результатам. Например:

In [58]:
# округление до двух знаков после запятой
print(round(2.545, 2))   # 2.54
print(round(2.555, 2))   # 2.56 - округление до четного
print(round(2.565, 2))   # 2.56
print(round(2.575, 2))   # 2.58
 
print(round(2.655, 2))   # 2.65 - округление не до четного
print(round(2.665, 2))   # 2.67
print(round(2.675, 2))   # 2.67

2.54
2.56
2.56
2.58
2.65
2.67
2.67


Подобно о проблеме можно почитать к документации.
Дополнительные материалы

# Поразрядные операции с числами

Особый класс операций представляют поразрядные операции. Они выполняются над отдельными разрядами чисел типа int. Но чтобы понять природу поразрядных операций, надо понимать что вообще представляет число в двоичном представление.
Двоичное представление чисел

При двоичной системе каждый разряд числа может иметь только два значения - 0 и 1. Например, 0 в десятичной системе также будет равен 0 в двоичной системе, а 1 в десятичной системе будет соответствовать 1 в двоичной системе. Следующее число в десятичной системе - 2 в двоичной системе будет соответствовать 10. То есть, когда мы к 1 прибавляем 1, то результатом будет 10. И так далее.

Например, 5 в двоичном представлении 101 и имеет три разряда. Для вывода десятичного числа в двоичной системе можно применять спецификатор 0b:

In [59]:
number = 5 # в двоичной форме 101
print(f"number = {number:0b}")  # number = 101

number = 101


Без указания спецификатора функция print() выводит число в десятичной системе.

При этом Python позволяет сразу определять число в двоичной форме. Для этого число в двоичной форме указывается после префикса 0b:

In [60]:
number = 0b101  # определяем число в двоичной форме
print(f"number = {number:0b}")  # number = 101
print(f"number = {number}")  # number = 5 - в десятичной системе

number = 101
number = 5


Еще несколько примеров сопоставления между двоичной и десятичной системами:

In [61]:
number1 = 1 # в двоичной системе 0b1
number2 = 2 # в двоичной системе 0b10
number3 = 3 # в двоичной системе 0b11
number4 = 4 # в двоичной системе 0b100
number5 = 5 # в двоичной системе 0b101
number6 = 6 # в двоичной системе 0b110

### Логические операции

Логические операции выполняются над отдельными разрядами числа. В Python есть следующие логические операции:

- `&` (логическое умножение): Умножение производится поразрядно, и если у обоих операндов значения разрядов равно 1, то операция возвращает 1, иначе возвращается число 0. Например:

In [62]:
x1 = 2  # 010
y1 = 5  # 101
z1 = x1 & y1
print(f"z1 = {z1}")   # z1 = 0
     
x2 = 4  # 100
y2 = 5  # 101
z2 = x2 & y2
print(f"z2 = {z2}")   # z2 = 4
print(f"z2 = {z2:0b}")  # z2 = 100

z1 = 0
z2 = 4
z2 = 100


В первом случае у нас два числа 2 и 5. 2 в двоичном виде представляет число 010, а 5 - 101. Поразрядно умножим числа (0*1, 1*0, 0*1) и в итоге получим 000.

Во втором случае у нас вместо двойки число 4, у которого в первом разряде 1, так же как и у числа 5, поэтому в итоге получим (1*1, 0*0, 0 *1) = 100, то есть число 4 в десятичном формате.

- `|` (логическое сложение): Похоже на логическое умножение, операция также производится по двоичным разрядам, но теперь возвращается единица, если хотя бы у одного числа в данном разряде имеется единица. Например:

In [ ]:
x1 = 2      # 010
y1 = 5      # 101
z1 = x1|y1  # 111
     
print(f"z1 = {z1}")     # z1 = 7
print(f"z1 = {z1:0b}")  # z1 = 111
     
x2 = 4          # 100
y2 = 5          # 101
z2 = x2 | y2    # 101
print(f"z2 = {z2}")     # z2 = 5
print(f"z2 = {z2:0b}")  # z2 = 101

- `^` (логическое исключающее ИЛИ): Если значения текущего разряда у обоих чисел разные, то возвращается 1, иначе возвращается 0. Также эту операцию называют XOR. Например:

In [63]:
x = 9       #  1001
y = 5       #  0101
z = x ^ y   #  1100
print(f"z = {z}")       # z = 12
print(f"z = {z:0b}")   # z = 1100

z = 12
z = 1100


Здесь число 9 в двоичной форме равно 1001. Число 5 равно 0101. Операция XOR дает следующий результат: 1^0, 0^1, 0^0, 1^1. Здесь мы видим, что первые два разряда чисел содержат разные значения, поэтому первые два разряда получат значение 1. А последние два разряда чисел содержат одинаковые значения, поэтому последние два разряда получат значение 0. Таким образом, мы получаем число 1100 или 12 в десятичной системе.

нередко данную операцию применяют для простого шифрования:

In [64]:
x = 45       # Значение, которое надо зашифровать - в двоичной форме 101101
key = 102    # Пусть это будет ключ - в двоичной форме 1100110
     
encrypt = x ^ key    # Результатом будет число 1001011 или 75
print(f"Зашифрованное число: {encrypt}")
     
decrypt = encrypt ^ key    # Результатом будет исходное число 45
print(f"Расшифрованное число: {decrypt}")

Зашифрованное число: 75
Расшифрованное число: 45


Также можно применять эту операцию для обмена значений чисел:

In [65]:
x = 9       #  1001
y = 5       #  0101
x = x ^ y
y = x ^ y
x = x ^ y 
     
print(f"x = {x}")       # x = 5
print(f"y = {y}")       # y = 9

x = 5
y = 9


- `~`(инверсия): Инвертирует число. Выражение ~x фактически аналогично -(x+1). Например:

In [66]:
x = 5
y = ~x;
print(f"y: {y}")  # -6

y: -6


### Операции сдвига

Операции сдвига также производятся над разрядами чисел. Сдвиг может происходить вправо и влево.

- `x<<y` - сдвигает число x влево на y разрядов. Например, 4<<1 сдвигает число 4 (которое в двоичном представлении 100) на один разряд влево, то есть в итоге получается 1000 или число 8 в десятичном представлении.

- `x>>y` - сдвигает число x вправо на y разрядов. Например, 16>>1 сдвигает число 16 (которое в двоичном представлении 10000) на один разряд вправо, то есть в итоге получается 1000 или число 8 в десятичном представлении.

Таким образом, если исходное число, которое надо сдвинуть в ту или другую строну, делится на два, то фактически получается умножение или деление на два. Поэтому подобную операцию можно использовать вместо непосредственного умножения или деления на два. Например:

In [67]:
a = 16 # в двоичной форме 10000
b = 2 
c = a << b  # Сдвиг числа 10000 влево на 2 разряда, равно 1000000 или 64 в десятичной системе
print(c)   #64
 
d = a >> b  #Сдвиг числа  10000  вправо на 2 разряда, равно 100 или 4 в десятичной системе
print(d)   #4

64
4


При этом числа, которые участвую в операциях, необязательно должны быть кратны 2::

In [68]:
a = 22 # в двоичной форме 10110
b = 2
c = a << b  # Сдвиг числа 10110 влево на 2 разряда, равно 1011000 или 88 в десятичной системе
print(c)   # 88
 
d = a >> b  # Сдвиг числа 10110 вправо на 2 разряда, равно 101 или 5 в десятичной системе
print(d)   # 5

88
5


# Условные выражения
Ряд операций представляют условные выражения. Все эти операции принимают два операнда и возвращают логическое значение, которое в Python представляет тип bool. Существует только два логических значения - True (выражение истинно) и False (выражение ложно).
Операции сравнения

Простейшие условные выражения представляют операции сравнения, которые сравнивают два значения. Python поддерживает следующие операции сравнения:

- `==` Возвращает True, если оба операнда равны. Иначе возвращает False.

- `!=` Возвращает True, если оба операнда НЕ равны. Иначе возвращает False.

- `>` (больше чем) Возвращает True, если первый операнд больше второго.

- `<` (меньше чем) Возвращает True, если первый операнд меньше второго.

- `>=` (больше или равно) Возвращает True, если первый операнд больше или равен второму.

- `<=` (меньше или равно) Возвращает True, если первый операнд меньше или равен второму.

Примеры операций сравнения:

In [70]:
a = 5
b = 6
result = 5 == 6  # сохраняем результат операции в переменную
print(result)  # False - 5 не равно 6
print(a != b)  # True
print(a > b)  # False - 5 меньше 6
print(a < b)  # True
 
bool1 = True
bool2 = False
print(bool1 == bool2)  # False - bool1 не равно bool2

False
True
False
True
False


Операции сравнения могут сравнивать различные объекты - строки, числа, логические значения, однако оба операнда операции должны представлять один и тот же тип.
Логические операции

Для создания составных условных выражений применяются логические операции. В Python имеются следующие логические операторы:

### Оператор and (логическое умножение) применяется к двум операндам:

`x and y`: Сначала оператор and оценивает выражение x, и если оно равно False, то возвращается его значение. Если оно равно True, то оценивается второй операнд - y и возвращается значение y.

In [71]:
age = 22
weight = 58
result = age > 21 and weight == 58
print(result)  # True

True


В данном случае оператор and сравнивает результаты двух выражений: age > 21 weight == 58. И если оба этих выражений возвращают True, то оператор and также возвращает True (формально возвращается значение последнего операнда).

Но операндами оператора and необязательно выступают значения True и False. Это могут быть любые значения. Например:

In [72]:
result = 4 and "w"
print(result)  # w, так как 4 равно True, поэтому возвращается значение последнего операнда
     
result = 0 and "w"
print(result)  # 0, так как 0 эквивалентно False

w
0


В данном случае число 0 и пустая строка "" расматриваются как False, все остальные числа и непустые строки эквивалентны True

### or (логическое сложение) также применяется к двум операндам:

`x or y`: Сначала оператор or оценивает выражение x, и если оно равно True, то возвращается его значение. Если оно равно False, то оценивается второй операнд - y и возвращается значение y. Например

In [73]:
age = 22
isMarried = False
result = age > 21 or isMarried
print(result)  # True, так как выражение age > 21 равно True

True


И также оператор or может применяться к любым значениям. Например:

In [74]:
result = 4 or "w"
print(result)  # 4, так как 4 эквивалентно True, поэтому возвращается значение первого операнда
     
result = 0 or "w"
print(result)  # w, так как 0 эквивалентно False, поэтому возвращается значение последнего операнда

4
w


### not (логическое отрицание)

Возвращает __True__, если выражение равно __False__

In [75]:
age = 22
isMarried = False
print(not age > 21)  # False
print(not isMarried)  # True
print(not 4)  # False
print(not 0)  # True

False
True
False
True


### Оператор in

Оператор in возвращает True если в некотором наборе значений есть определенное значение. Он имеет следующую форму:

> значение in набор_значений

Например, строка представляет набор символов. И с помощью оператора in мы можем проверить, есть ли в ней какая-нибудь подстрока:

In [76]:
message = "hello world!"
hello = "hello"
print(hello in message)  # True - подстрока hello есть в строке "hello world!"
 
gold = "gold"
print(gold in message)  # False - подстроки "gold" нет в строке "hello world!"

True
False


Если нам надо наоборот проверить, нет ли в наборе значений какого-либо значения, то мы може использовать модификацию оператора - not in. Она возвращает True, если в наборе значений НЕТ определенного значения:

In [77]:
message = "hello world!"
hello = "hello"
print(hello not in message)  # False
 
gold = "gold"
print(gold not in message)  # True

False
True


# Условная конструкция if
Условные конструкции используют условные выражения и в зависимости от их значения направляют выполнение программы по одному из путей. Одна из таких конструкций - это конструкция __if__. Она имеет следующее формальное определение:
```
if логическое_выражение:
    инструкции
[elif логическое выражение:
    инструкции]
[else: 
    инструкции]
```
В самом простом виде после ключевого слова __if__ идет логическое выражение. И если это логическое выражение возвращает True, то выполняется последующий блок инструкций, каждая из которых должна начинаться с новой строки и должна иметь отступы от начала выражения if (отступ желательно делать в 4 пробела или то количество пробелов, которое кратно 4):

In [78]:
language = "english"
if language == "english":
    print("Hello")
print("End")

Hello
End


Поскольку в данном случае значение переменной language равно "english", то будет выполняться блок if, который содержит только одну инструкцию - print("Hello"). В итоге консоль выведет следующие строки:
```
Hello
End
```
Обратите внимание в коде на последнюю строку, которая выводит сообщение "End". Она не имеет отступов от начала строки, поэтому она не принадлежит к блоку if и будет выполняться в любом случае, даже если выражение в конструкции if возвратит False.

Но если бы мы поставили бы отступы, то она также принадлежала бы к конструкции if:

In [79]:
language = "english"
if language == "english":
    print("Hello")
    print("End")
    

Hello
End


### Блок else

Если вдруг нам надо определить альтернативное решение на тот случай, если выражение в __if__ возвратит False, то мы можем использовать блок else:

In [80]:
language = "russian"
if language == "english":
    print("Hello")
else:
    print("Привет")
print("End")

Привет
End


Если выражение `language == "english"` возвращает True, то выполняется блок if, иначе выполняется блок else. И поскольку в данном случае условие `language == "english"` возвращает False, то будут выполняться инструкция из блока else.

Причем инструкции блока else также должны имет отступы от начала строки. Например, в примере выше print("End") не имеет отступа, поэтому она не входит в блок else и будет выполнятьься вне зависимости, чему равно условие `language == "english"`. То есть консоль нам выведет следующие строки:
```
Привет
End
```
Блок else также может иметь несколько инструкций, которые должны иметь отступ от начала строки:

In [83]:
language = "russian"
if language == "english":
    print("Hello")
    print("World")
else:
    print("Привет")
    print("мир")


Привет
мир


Если необходимо ввести несколько альтернативных условий, то можно использовать дополнительные блоки elif, после которого идет блок инструкций.

In [84]:
language = "german"
if language == "english":
    print("Hello")
    print("World")
elif language == "german":
    print("Hallo")
    print("Welt")
else:
    print("Привет")
    print("мир")

Hallo
Welt


Сначала Python проверяет выражение __if__. Если оно равно True, то выполняются инструкции из блока if. Если это условие возвращает False, то Python проверяет выражение из __elif__.

Если выражение после __elif__ равно True, то выполняются инструкции из блока __elif__. Но если оно равно False то выполняются инструкции из блока __else__

При необходимости можно определить несколько блоков __elif__ для разных условий. Например:

In [85]:
language = "german"
if language == "english":
    print("Hello")
elif language == "german":
    print("Hallo")
elif language == "french":
    print("Salut")
else:
    print("Привет")
    

Hallo


### Вложенные конструкции if

Конструкция __if__ в свою очередь сама может иметь вложенные конструкции __if__:

In [86]:
language = "english"
daytime = "morning"
if language == "english":
    print("English")
    if daytime == "morning":
        print("Good morning")
    else:
        print("Good evening")

English
Good morning


Здесь конструкция __if__ содержит вложенную конструкцию __if/else__. То есть если переменная language равна "english", тогда вложенная конструкция __if/else__ дополнительно проверяет значение переменной daytime - равна ли она строке "morning" ли нет. И в данном случае мы получим следующий консольный вывод:
```
English
Good morning
```
Стоит учитывать, что вложенные выражения __if__ также должны начинаться с отступов, а инструкции во вложенных конструкциях также должны иметь отступы. Отступы, расставленные не должным образом, могут изменить логику программы. Так, предыдущий пример НЕ аналогичен следующему:

In [87]:
language = "english"
daytime = "morning"
if language == "english":
    print("English")
if daytime == "morning":
    print("Good morning")
else:
     print("Good evening")

English
Good morning


Подобным образом можно размещать вложенные конструкции if/elif/else в блоках elif и else:

In [88]:
language = "russian"
daytime = "morning"
if language == "english":
    if daytime == "morning":
        print("Good morning")
    else:
        print("Good evening")
else:
    if daytime == "morning":
        print("Доброе утро")
    else:
        print("Добрый вечер")

Доброе утро


# Циклы

Циклы позволяют выполнять некоторое действие в зависимости от соблюдения некоторого условия. В языке Python есть следующие типы циклов:

- __while__

- __for__

### Цикл while

Цикл while проверяет истинность некоторого условия, и если условие истинно, то выполняет инструкции цикла. Он имеет следующее формальное определение:

```
while условное_выражение:
   инструкции
```
После ключевого слова __while__ указывается условное выражение, и пока это выражение возвращает значение True, будет выполняться блок инструкций, который идет далее.

Все инструкции, которые относятся к циклу while, располагаются на последующих строках и должны иметь отступ от начала ключевого слова while.

In [1]:
number = 1
 
while number < 5:
    print(f"number = {number}")
    number += 1
print("Работа программы завершена")

number = 1
number = 2
number = 3
number = 4
Работа программы завершена


В данном случае цикл while будет выполняться, пока переменная number меньше 5.

Сам блок цикла состоит из двух инструкций:

In [2]:
print(f"number = {number}")
number += 1

number = 5


Обратите внимание, что они имеют отступы от начала оператора while - в данном случае от начала строки. Благодаря этому Python может определить, что они принадлежат циклу. В самом цикле сначала выводится значение переменной number, а потом ей присваивается новое значение.

Также обратите внимание, что последняя инструкция print("Работа программы завершена") не имеет отступов от начала строки, поэтому она не входит в цикл while.

Весь процесс цикла можно представить следующим образом:

1) Сначала проверяется значение переменной number - меньше ли оно 5. И поскольку вначале переменная равна 1, то это условие возвращает True, и поэтому выполняются инструкции цикла

2) Инструкции цикла выводят на консоль строку number = 1. И далее значение переменной number увеличивается на единицу - теперь она равна 2. Однократное выполнение блока инструкций цикла называется итерацией. То есть таким образом, в цикле выполняется первая итерация.

3) Снова проверяется условие number < 5. Оно по прежнему равно True, так как number = 2, поэтому выполняются инструкции цикла

4) Инструкции цикла выводят на консоль строку number = 2. И далее значение переменной number опять увеличивается на единицу - теперь она равна 3. Таким образом, выполняется вторая итерация.

6) Опять проверяется условие number < 5. Оно по прежнему равно True, так как number = 3, поэтому выполняются инструкции цикла

7) Инструкции цикла выводят на консоль строку number = 3. И далее значение переменной number опять увеличивается на единицу - теперь она равна 4. То есть выполняется третья итерация.

8) Снова проверяется условие number < 5. Оно по прежнему равно True, так как number = 4, поэтому выполняются инструкции цикла

9) Инструкции цикла выводят на консоль строку number = 4. И далее значение переменной number опять увеличивается на единицу - теперь она равна 5. То есть выполняется четвертая итерация.

10) И вновь проверяется условие number < 5. Но теперь оно равно False, так как number = 5, поэтому выполняются выход из цикла. Все цикл - завершился. Дальше уже выполняются действия, которые определены после цикла. Таким образом, данный цикл произведет четыре прохода или четыре итерации

В итоге при выполнении кода мы получим следующий консольный вывод:
```
number = 1
number = 2
number = 3
number = 4
Работа программы завершена

```
Для цикла while также можно определить дополнительный блок else, инструкции которого выполняются, когда условие равно False:

In [3]:
number = 1
 
while number < 5:
    print(f"number = {number}")
    number += 1
else:
    print(f"number = {number}. Работа цикла завершена")
print("Работа программы завершена")

number = 1
number = 2
number = 3
number = 4
number = 5. Работа цикла завершена
Работа программы завершена


То есть в данном случае сначала проверяется условие и выполняются инструкции while. Затем, когда условие становится равным False, выполняются инструкции из блока else. Обратите внимание, что инструкции из блока else также имеют отступы от начала конструкции цикла. 

Блок else может быть полезен, если условие изначально равно False, и мы можем выполнить некоторые действия по этому поводу:

In [4]:
number = 10
 
while number < 5:
    print(f"number = {number}")
    number += 1
else:
    print(f"number = {number}. Работа цикла завершена")
print("Работа программы завершена")

number = 10. Работа цикла завершена
Работа программы завершена


В данном случае условие number < 5 изначально равно False, поэтому цикл не выполняет ни одной итерации и сразу переходит в блоку else.

### Цикл for

Другой тип циклов представляет конструкция __for__. Этот цикл пробегается по набору значений, помещает каждое значение в переменную, и затем в цикле мы можем с этой переменной производить различные действия. Формальное определение цикла for:
```
for переменная in набор_значений:
    инструкции
```
После ключевого слова __for__ идет название переменной, в которую будут помещаться значения. Затем после оператора __in__ указывается набор значений и двоеточие.

А со следующей строки располагается блок инструкций цикла, которые также должны иметь отступы от начала цикла.

При выполнении цикла Python последовательно получает все значения из набора и передает их переменную. Когда все значения из набора будут перебраны, цикл завершает свою работу.

В качестве набора значений, например, можно рассматривать строку, которая по сути представляет набор символов. Посмотрим на примере:

In [5]:
message = "Hello"
 
for c in message:
    print(c)

H
e
l
l
o


В цикле определяется переменную c, после оператора in в качестве перебираемого набора указана переменная message, которая хранит строку "Hello". В итоге цикл for будет перебираеть последовательно все символы из строки message и помещать их в переменную c. Блок самого цикла состоит из одной инструкции, которая выводит значение переменной с на консоль. Консольный вывод программы:

Цикл __for__ также может иметь дополнительный блок __else__, который выполняется после завершения цикла:

In [6]:
message = "Hello"
for c in message:
    print(c)
else:
    print(f"Последний символ: {c}. Цикл завершен");
print("Работа программы завершена")  # инструкция не имеет отступа, поэтому не относится к else

H
e
l
l
o
Последний символ: o. Цикл завершен
Работа программы завершена



Стоит отметить, что блок else имеет доступ ко всем переменным, которые определены в цикле for.

### Вложенные циклы

Одни циклы внутри себя могут содержать другие циклы. Рассмотрим на примере вывода таблицы умножения:

In [7]:
i = 1
j = 1
while i < 10:
    while j < 10:
        print(i * j, end="\t")
        j += 1
    print("\n")
    j = 1
    i += 1

1	2	3	4	5	6	7	8	9	

2	4	6	8	10	12	14	16	18	

3	6	9	12	15	18	21	24	27	

4	8	12	16	20	24	28	32	36	

5	10	15	20	25	30	35	40	45	

6	12	18	24	30	36	42	48	54	

7	14	21	28	35	42	49	56	63	

8	16	24	32	40	48	56	64	72	

9	18	27	36	45	54	63	72	81	



Внешний цикл while i < 10: срабатывает 9 раз пока переменная i не станет равна 10. Внутри этого цикла срабатывает внутренний цикл while j < 10:. Внутренний цикл также срабатывает 9 раз пока переменная j не станет равна 10. Причем все 9 итераций внутреннего цикла срабатывают в рамках одной итерации внешнего цикла.

В каждой итерации внутреннего цикла на консоль будет выводится произведение чисел i и j. Затем значение переменной j увеличивается на единицу. Когда внутренний цикл закончил работу, значений переменной j сбрасывается в 1, а значение переменной i увеличивается на единицу и происходит переход к следующей итерации внешнего цикла. И все повторяется, пока переменная i не станет равна 10. Соответственно внутренний цикл сработает всего 81 раз для всех итераций.

Подобным образом можно определять вложенные циклы for:

In [8]:
for c1 in  "ab":
    for c2 in "ba":
        print(f"{c1}{c2}")

ab
aa
bb
ba


В данном случае внешний цикл проходит по строке "ab" и каждый символ помещает в переменную c1. Внутренний цикл проходит по строке "ba", помещает каждый символ строки в переменную c2 и выводит сочетание обоих символов на консоль. То есть в итоге мы получим все возможные сочетания символов a и b:

### Выход из цикла. break и continue

Для управления циклом мы можем использовать специальные операторы break и continue. Оператор break осуществляет выход из цикла. А оператор continue выполняет переход к следующей итерации цикла.

Оператор break может использоваться, если в цикле образуются условия, которые несовместимы с его дальнейшим выполнением. Рассмотрим следующий пример:

In [9]:
number = 0
while number < 5:
    number += 1
    if number == 3 :    # если number = 3, выходим из цикла
        break
    print(f"number = {number}")

number = 1
number = 2


Здесь цикл while проверяет условие number < 5. И пока number не равно 5, предполагается, что значение number будет выводиться на консоль. Однако внутри цикла также проверяется другое условие: if number == 3. То есть, если значение number равно 3, то с помощью оператора break выходим из цикла. И в итоге мы получим следующий консольный вывод:

В отличие от оператора break оператор continue выполняет переход к следующей итерации цикла без его завершения. Например, в предыдущем примере заменим break на continue:

In [10]:
number = 0
while number < 5:
    number += 1
    if number == 3 :    # если number = 3, переходим к новой итерации цикла
        continue
    print(f"number = {number}")

number = 1
number = 2
number = 4
number = 5


И в этом случае если значение переменной number равно 3, последующие инструкции после оператора continue не будут выполняться

# Функции


Функции представляют блок кода, который выполняет определенную задачу и который можно повторно использовать в других частях программы. В предыдущих статьях уже использовались функции. В частности, функция print(), которая выводит некоторое значение на консоль. Python имеет множество встроенных функций и позволяет определять свои функции. Формальное определение функции:
```
def имя_функции ([параметры]):
    инструкции
```
Определение функции начинается с выражения def, которое состоит из имени функции, набора скобок с параметрами и двоеточия. Параметры в скобках необязательны. А со следующей строки идет блок инструкций, которые выполняет функция. Все инструкции функции имеют отступы от начала строки.

Например, определение простейшей функции:

In [11]:
def say_hello():
    print("Hello")

Функция называется say_hello. Она не имеет параметров и содержит одну единственную инструкцию, которая выводит на консоль строку "Hello".

Обратите внимание, что инструкции функции должны иметь отступы от начала функции. Например:

In [12]:
def say_hello():
    print("Hello")
 
 
print("Bye")

Bye


Здесь инструкция print("Bye") не имеет отступов от начала функции say_hello и поэтому в эту функцию не входит. Обычно между определением функции и остальными инструкциями, которые не входят в функцию, располагаются две пустых строки.

Для вызова функции указывается имя функции, после которого в скобках идет передача значений для всех ее параметров:
```
имя_функции ([параметры])
```
Например, определим и вызовем функцию:

In [13]:
def say_hello():    # определение функции say_hello
    print("Hello")
 
 
say_hello()         # вызов функции say_hello
say_hello()
say_hello()

Hello
Hello
Hello


Здесь три раза подряд вызывается функция say_hello. Обратите внимание, что функция сначала определяется, а потом вызывается.

Если функция имеет одну инструкцию, то ее можно разместить на одной строке с остальным определением функции:

In [14]:
def say_hello(): print("Hello")

say_hello()

Hello


Подобным образом можно определять и вызывать и другие функции. Например, определим и выполним несколько функций:

In [15]:
def say_hello():
    print("Hello")
 
 
def say_goodbye():
    print("Good Bye")
 
 
say_hello()
say_goodbye()

Hello
Good Bye


### Локальные функции

Одни функции могут определяться внутри других функций - внутренние функции еще называют локальными. Локальные функции можно использовать только внутри той функции, в которой они определены. Например:

In [16]:
def print_messages():
    # определение локальных функций
    def say_hello(): print("Hello")
    def say_goodbye(): print("Good Bye")
    # вызов локальных функций
    say_hello()
    say_goodbye()

# Вызов функции print_messages
print_messages()
 
#say_hello() # вне функции print_messages функция say_hello не доступна

Hello
Good Bye


Здесь функции say_hello() и say_goodbye() определены внутри функции print_messages() и поэтому по отношению к ней являются локальными. Соответственно они могут использоваться только внутри функции print_messages()
Организация программы и функция main

В программе может быть определено множество функций. И чтобы всех их упорядочить, одним из способов их организации является добавление специальной функции (обычно называется main), в которой потом уже вызываются другие функции:

In [17]:
def main():
    say_hello()
    say_goodbye()

def say_hello():
    print("Hello")

def say_goodbye():
    print("Good Bye")

# Вызов функции main
main()

Hello
Good Bye


# Параметры функции
Функция может принимать параметры. Через параметры в функцию можно передавать данные. Банальный пример - функция print(), которая с помощью параметра принимает значение, выводимое на консоль.

Теперь определим и используем свою функцию с параметрами:

In [18]:
def say_hello(name):
    print(f"Hello, {name}")

say_hello("Tom")
say_hello("Bob")
say_hello("Alice")

Hello, Tom
Hello, Bob
Hello, Alice


Функция say_hello имеет параметр name, и при вызове функции мы можем передать этому параметру какой-либо значение. Внутри функции мы можем использовать параметр как обычную переменную, например, вывести значение этого параметра на консоль функцией print. Так, в выражении:

In [19]:
say_hello("Tom")

Hello, Tom


Строка "Tom" будет передаваться параметру name

При вызове функции значения передаются параметрам по позиции. Например, определим и вызовем функцию с несколькими параметрами:

In [20]:
def print_person(name, age):
    print(f"Name: {name}")
    print(f"Age: {age}")
 
 
print_person("Tom", 37)

Name: Tom
Age: 37


Здесь функция print_person принимает два параметра: name и age. При вызове функции:

In [21]:
print_person("Tom", 37)

Name: Tom
Age: 37


Первое значение - "Tom" передается первому параметру, то есть параметру name. Второе значение - 37 передается второму параметру - age. И внутри функции значения параметров выводятся на консоль:
```
Name: Tom
Age: 37
```
### Значения по умолчанию

Некоторые параметры функции мы можем сделать необязательными, указав для них значения по умолчанию при определении функции. Например:

In [22]:
def say_hello(name="Tom"):
    print(f"Hello, {name}")
 
 
say_hello()         # здесь параметр name будет иметь значение "Tom"
say_hello("Bob")    # здесь name = "Bob"

Hello, Tom
Hello, Bob


Здесь параметр name является необязательным. И если мы не передаем при вызове функции для него значение, то применяется значение по умолчанию, то есть строка "Tom". Консольный вывод данной программы:

Если функция имеет несколько параметров, то необязательные параметры должны идти после обязательных. Например:

In [24]:
def print_person(name, age = 18):
    print(f"Name: {name}  Age: {age}")
 
 
print_person("Bob")
print_person("Tom", 37)

Name: Bob  Age: 18
Name: Tom  Age: 37


Здесь параметр age является необязательным и по умолчанию имеет значение 18. Перед ним расположен обязательный параметр name. Поэтому при вызове функции мы можем не передавать значение параметру age, но параметру name передать значение необходимо.

При необходимости мы можем сделать все параметры необязательными:

In [23]:
def print_person(name = "Tom", age = 18):
    print(f"Name: {name}  Age: {age}")
 
 
print_person()              # Name: Tom  Age: 18
print_person("Bob")         # Name: Bob  Age: 18
print_person("Sam", 37)     # Name: Sam  Age: 37

Name: Tom  Age: 18
Name: Bob  Age: 18
Name: Sam  Age: 37


### Передача значений параметрам по имени. Именованные параметры

В примерах выше при вызове функции значения передаются параметрами функции по позиции. Но также можно передавать значения параметрам по имени. Для этого при вызове функции указывается имя параметра и ему присваивается значение:

In [25]:
def print_person(name, age):
    print(f"Name: {name}  Age: {age}")
 
 
print_person(age = 22, name = "Tom")

Name: Tom  Age: 22


В данном случае значения параметрам age и name передаются по имени. И несмотря на то, что параметр name идет первым в определении функции, мы можем при вызове функции написать print_person(age = 22, name = "Tom") и таким образом передать число 22 параметру age, а строку "Tom" параметру name.

Символ `*`позволяет установить, какие параметры будут именнованными - то есть такие параметры, которым можно передать значения только по имени. Все параметры, которые располагаются справа от символа `*`, получают значения только по имени:

In [26]:
def print_person(name, *,  age, company):
    print(f"Name: {name}  Age: {age}  Company: {company}")
 
 
print_person("Bob", age = 41, company ="Microsoft")    # Name: Bob  Age: 41  company: Microsoft

Name: Bob  Age: 41  Company: Microsoft


В данном случае параметры age и company являются именнованными.

Можно сделать все параметры именнованными, поставив перед списком параметров символ *:

In [27]:
def print_person(*,  name, age, company):
    print(f"Name: {name}  Age: {age}  Company: {company}")

Если наоборот надо определить параметры, которым можно передавать значения только по позиции, то есть позиционные параметры, то можно использовать символ /: все параметры, которые идут до символа / , являются позиционными и могут получать значения только по позиции

In [28]:
def print_person(name, /, age, company="Microsoft"):
    print(f"Name: {name}  Age: {age}  Company: {company}")
 
 
print_person("Tom", company="JetBrains", age = 24)     # Name: Tom  Age: 24  company: JetBrains
print_person("Bob", 41)                 # Name: Bob  Age: 41  company: Microsoft

Name: Tom  Age: 24  Company: JetBrains
Name: Bob  Age: 41  Company: Microsoft


В данном случае параметр name является позиционным.

Для одной функции можно определять одновременно позиционные и именнованные параметры.

In [29]:
def print_person(name, /,  age = 18, *, company):
    print(f"Name: {name}  Age: {age}  Company: {company}")
 
 
print_person("Sam", company ="Google")               # Name: Sam  Age: 18  company: Google
print_person("Tom", 37, company ="JetBrains")        # Name: Tom  Age: 37  company: JetBrains
print_person("Bob", company ="Microsoft", age = 42)  # Name: Bob  Age: 42  company: Microsoft

Name: Sam  Age: 18  Company: Google
Name: Tom  Age: 37  Company: JetBrains
Name: Bob  Age: 42  Company: Microsoft


В данном случае параметр name располагается слева от символа /, поэтому является позиционным и обязательным - ему можно передать значение только по позиции.

Параметр company является именнованным, так как располагается справа от символа `*`. Параметр age может получать значение по имени и по позиции.

### Неопределенное количество параметров

С помощью символа звездочки можно определить параметр, через который можно передавать неопределенное количество значений. Это может быть полезно, когда мы хотим, чтобы функция получала несколько значений, но мы точно не знаем, сколько именно. Например, определим функцию подсчета суммы чисел:

In [30]:
def sum(*numbers):
    result = 0
    for n in numbers:
        result += n
    print(f"sum = {result}")
 
 
sum(1, 2, 3, 4, 5)      # sum = 15
sum(3, 4, 5, 6)         # sum = 18

sum = 15
sum = 18


В данном случае функция __sum__ принимает один параметр - `*numbers`, но звездочка перед названием параметра указывает, что фактически на место этого параметра мы можем передать неопределенное количество значений или набор значений. В самой функции с помощью цикла for можно пройтись по этому набору, получить каждое значение из этого набора в переменную n и произвести с ним какие-нибудь действия. Например, в данном случае вычисляется сумма переданных чисел.

# Оператор return и возвращение результата из функции
### Возвращение результата

Функция может возвращать результат. Для этого в функции используется оператор return, после которого указывается возвращаемое значение:
```
def имя_функции ([параметры]):
    инструкции
    return возвращаемое_значение
```
Определим простейшую функцию, которая возвращает значение:

In [31]:
def get_message():
    return "Hello METANIT.COM"

Здесь после оператора return идет строка "Hello METANIT.COM" - это значение и будет возвращать функция get_message().

Затем это результат функции можно присвоить переменной или использовать как обычное значение:

In [32]:
def get_message():
    return "Hello METANIT.COM"
 
message = get_message()  # получаем результат функции get_message в переменную message
print(message)          # Hello METANIT.COM
 
# можно напрямую передать результат функции get_message
print(get_message())    # Hello METANIT.COM

Hello METANIT.COM
Hello METANIT.COM


После оператора __return__ может идти и сложное вычислямое выражение, резлуьтат которого будет возвращаться из функции. Например, определим функцию, которая увеличивает число в два раза:

In [33]:
def double(number):
    return 2 * number

Здесь функция double будет возвращать результат выражения 2 * number:

In [34]:
def double(number):
    return 2 * number
 
result1 = double(4)     # result1 = 8
result2 = double(5)     # result2 = 10
print(f"result1 = {result1}")   # result1 = 8
print(f"result2 = {result2}")   # result2 = 10

result1 = 8
result2 = 10


Или другой пример - получение суммы чисел:

In [35]:
def sum(a, b):
    return a + b
 
result = sum(4, 6)                  # result = 0
print(f"sum(4, 6) = {result}")      # sum(4, 6) = 10
print(f"sum(3, 5) = {sum(3, 5)}")   # sum(3, 5) = 8

sum(4, 6) = 10
sum(3, 5) = 8


### Выход из функции

Оператор __return__ не только возвращает значение, но и производит выход из функции. Поэтому он должен определяться после остальных инструкций. Например:

In [36]:
def get_message():
    return "Hello METANIT.COM"
    print("End of the function")

print(get_message())

Hello METANIT.COM


С точки зрения синтаксиса данная функция корректна, однако ее инструкция print("End of the function") не имеет смысла - она никогда не выполнится, так как до ее выполнения оператор return возвратит значение и произведет выход из функции.

Однако мы можем использовать оператор return и в таких функциях, которые не возвращают никакого значения. В этом случае после оператора return не ставится никакого возвращаемого значения. Типичная ситуация - в зависимости от опеределенных условий произвести выход из функции:

In [37]:
def print_person(name, age):
    if age > 120 or age < 1:
        print("Invalid age")
        return
    print(f"Name: {name}  Age: {age}")
 
 
print_person("Tom", 22)
print_person("Bob", -102)

Name: Tom  Age: 22
Invalid age


Здесь функция print_person в качестве параметров принимает имя и возраст пользователя. Однако в функции вначале мы проверяем, соответствует ли возраст некоторому диапазону (меньше 120 и больше 0). Если возраст находится вне этого диапазона, то выводим сообщение о недопустимом возрасте и с помощью оператора return выходим из функции. После этого функция заканчивает свою работу.

Однако если возраст корректен, то выводим информацию о пользователе на консоль

# Функция как тип, параметр и результат другой функции
### Функция как тип

В Python функция фактически представляет отдельный тип. Так мы можем присвоить переменной какую-нибудь функцию и затем, используя переменную, вызывать данную функцию. Например:

In [38]:
def say_hello(): print("Hello")
def say_goodbye(): print("Good Bye")

message = say_hello
message()       # Hello
message = say_goodbye
message()       # Good Bye

Hello
Good Bye


В данном случае переменной message присваивается одна из функций. Сначала ей передается функция say_hello():

In [39]:
message = say_hello

После этого переменная message будет указывать на данную функцию, то есть фактически представлять функцию say_hello. А это значит, что мы можем вызывать переменную message как обычную функцию:

In [40]:
message()       # Hello

Hello


Фактически это приведет к выполнению функции say_hello, и на консоль будет выведена строка "Hello". Затем подобным образом мы можем передать переменной message другую функцию и вызвать ее.

Подобным образом можно через переменную вызывать функцию с параметрами и возвращать ее результат:

In [41]:
def sum(a, b): return a + b
def multiply(a, b): return a * b
 
operation = sum
result = operation(5, 6)
print(result)   # 11
 
operation = multiply
print(operation(5, 6))      # 30

11
30


### Функция как параметр функции

Поскольку функция в Python может представлять такое же значение как строка или число, соответственно мы можем передать ее в качестве параметра в другую функцию. Например, определим функцию, которая выводит на консоль результат некоторой операции:

In [42]:
def do_operation(a, b, operation):
    result = operation(a, b)
    print(f"result = {result}")

def sum(a, b): return a + b
def multiply(a, b): return a * b
 
do_operation(5, 4, sum)         # result = 9
do_operation(5, 4, multiply)   # result = 20

result = 9
result = 20


В данном случае функция do_operation имеет три параметра, причем третий параметр, как предполагается, будет представлять функцию, которая принимает два параметра и возвращает некоторый результат. Иными словами третий параметр - operation представляет некоторую операцию, но на момент определения функции do_operation мы точно не знаем, что это будет за операция. Мы только знаем, что она принимает два параметр и возвращает какой-то результат, который потом выводится на консоль.

При вызове функции do_operation мы сможем передать в качестве третьего параметра другую функцию, например, функцию sum:

In [43]:
do_operation(5, 4, sum)

result = 9


То есть в данном случае параметр operation фактически будет представлять функцию sum и будет возвращать сумму дву чисел.

Затем аналогичным образов в вызов функции do_operation можно передать третьему параметру другую функцию - multiply, которая выполнит умножение чисел:

In [44]:
do_operation(5, 4, multiply)   # result = 20

result = 20


Таким образом, более гибкие по функциональности функции, которые через параметры принимают другие функции.
Функция как результат функции

Также одна функция в Python может возвращать другую функцию. Например, определим функцию, которая в зависимости от значения параметра возвращает ту или иную операцию:

In [45]:
def sum(a, b): return a + b
def subtract(a, b): return a - b
def multiply(a, b): return a * b
 
def select_operation(choice):
    if choice == 1:
        return sum
    elif choice == 2:
        return subtract
    else:
        return multiply
 
 
operation = select_operation(1)     # operation = sum
print(operation(10, 6))             # 16
 
operation = select_operation(2)     # operation = subtract
print(operation(10, 6))             # 4
 
operation = select_operation(3)     # operation = multiply
print(operation(10, 6))             # 60

16
4
60


В данном случае функция select_operation в зависимости от значения параметра choice возвращает одну из трех функций - sum, subtract и multiply. Затем мы мы можем получить результат функции select_operation в переменную operation:

In [46]:
operation = select_operation(1)

Так, в данном случае в функцию select_operation передается число 1, соответственно она будет возвращать функцию sum. Поэтому переменная operation фактически будет указывать на функцию sum, которая выполняет сложение двух чисел:

In [47]:
print(operation(10, 6))             # 16 - фактически равно sum(10, 6)

16


Аналогичным образом можно получить и выполнить другие функции.

# Лямбда-выражения

Лямбда-выражения в языке Python представляют небольшие анонимные функции, которые определяются с помощью оператора __lambda__. Формальное определение лямбда-выражения:
```
lambda [параметры] : инструкция
```
Определим простейшее лямбда-выражение:

In [48]:
message = lambda: print("hello")
 
message()   # hello

hello


Здесь лямбда-выражение присваивается переменной message. Это лямбда-выражение не имеет параметров, ничего не возвращает и просто выводит строку "hello" на консоль. И через переменную message мы можем вызвать это лямбда-выражение как обычную функцию. Фактически оно аналогично следующей функции:

In [49]:
def message(): 
    print("hello")

Если лямбда-выражение имеет параметры, то они определяются после ключевого слова lambda. Если лямбда-выражение возвращает какой-то результат, то он указывается после двоеточия. Например, определим лямбда-выражение, которое возвращает квадрат числа:

In [50]:
square = lambda n: n * n
 
print(square(4))    # 16
print(square(5))    # 25

16
25


В данном случае лямбда-выражение принимает один параметр - n. Справа от двоеточия идет возвращаемое значение - n* n. Это лямбда-выражение аналогично следующей функции:

In [51]:
def square2(n): return n * n

Аналогичным образом можно создавать лямбда-выражения, которые принимают несколько параметров:

In [52]:
sum = lambda a, b: a + b
 
print(sum(4, 5))    # 9
print(sum(5, 6))    # 11

9
11


Хотя лямбда-выражения позволяют немного сократить определения функций, тем не менее они ограничены тем, что они могут выполнять только одно выражение. Однако они могут быть довольно удобны в тех случаях, когда необходимо использовать функцию для передачи в качестве параметра или возвращения в другой функции. Например, передача лямбда-выражения в качестве параметра:

In [53]:
def do_operation(a, b, operation):
    result = operation(a, b)
    print(f"result = {result}")

do_operation(5, 4, lambda a, b: a + b)  # result = 9
do_operation(5, 4, lambda a, b: a * b)  # result = 20

result = 9
result = 20


В данном случае нам нет необходимости определять функции, чтобы передать их в качестве параметра, как в прошлой статье.

То же самое касается и возвращение лямбда-выражений из функций:

In [54]:
def select_operation(choice):
    if choice == 1:
        return lambda a, b: a + b
    elif choice == 2:
        return lambda a, b: a - b
    else:
        return lambda a, b: a * b
 
 
operation = select_operation(1)  # operation = sum
print(operation(10, 6))  # 16
 
operation = select_operation(2)  # operation = subtract
print(operation(10, 6))  # 4
 
operation = select_operation(3)  # operation = multiply
print(operation(10, 6))  # 60

16
4
60


# Преобразование типов

В операциях с данными могут применяться значения различных типов. Например, складываются число типа int и число типа float:

In [55]:
a = 2       # число int
b = 2.5     # число float
c = a + b
print(c)    # 4.5

4.5


В данном случае никакой ошибки не будет. Однако Python не всегда может автоматически производить операции, в которых участвуют данные разных типов. Рассмотрим, какие в данном случае действуют правила.
Неявные преобразования

Оба числа в арифметических операциях должны представлять один и тот же тип. Если же два операнда операции представляют разные типы данных, то Python пытается автоматически выполнить преобразования к одному из типов в соответствии со следующми правилами:

- Если один из операндов операции представляет комплексное число (тип complex), то другой операнд также преобразуется к типу complex.

- Иначе, если один из операндов представляет тип float, то второй операнд также преобразуется к типу float. Собственно так и произошло в примере выше, где значение переменной a было преобразовано в тип float

- Иначе, оба операнда должны представлять тип int, и в этом случае преобазование не требуется

### Явные преобразования

Но в некоторых случаях возникает необходимость вручную выполнить преобразование типов. Например, пусть у нас будет следующий код:

In [56]:
a = "2"
b = 3
c = a + b

TypeError: can only concatenate str (not "int") to str

Мы ожидаем, что "2" + 3 будет равно 5. Однако этот код сгенерирует исключение, так как первое число на самом деле представляет строку. И мы увидим при выполнении кода что-то наподобие:
```
Traceback (most recent call last):
  File "/Users/eugene/PycharmProjects/HelloApp/main.py", line 3, in 
    c = a + b
TypeError: can only concatenate str (not "int") to str
```
Для преобразования типов Python предоставляет ряд встроенных функций:

- int(): преобразует значение в целое число

- float(): преобразует значение в число с плавающей точкой

- str(): преобразует значение в строку

### int

Так, в предыдущем примере преобазуем строку в число с помощью функции int():

In [57]:
a = "2"
b = 3
c = int(a) + b
print(c)    # 5

5


Примеры преобразований с помощью int():

In [58]:
a = int(15)     # a = 15
b = int(3.7)    # b = 3
c = int("4")    # c = 4
e = int(False)    # e = 0
f = int(True)     # f = 1

Однако если значение не может быть преобразовано, то функция int выдаст ошибку ValueError: invalid literal for int() with base 10:

In [59]:
b = int("a1c")    # Ошибка
c = int("4.7")    # Ошибка

ValueError: invalid literal for int() with base 10: 'a1c'

### float

Аналогичным образом действует функция float(), которая преобразует в число с плавающей точкой.

In [60]:
a = "2.7"
b = 3
c = float(a) + b
print(c) # 5.7

5.7


Примеры преобразований с помощью float():

In [61]:
a = float(15)       # a = 15.0
b = float(3.7)      # b = 3.7
c = float("4.7")    # c = 4.7
d = float("5")      # d = 5.0
e = float(False)    # e = 0.0
f = float(True)     # f = 1.0

Но опять же не все значения могут автоматически преобразованы в float. Так, в следующем случае Python сгенерирует ошибку:

In [62]:
d = float("abc")  # Ошибка

ValueError: could not convert string to float: 'abc'

### str

Функция str() преобразует значение в строку:

In [63]:
a = str(False)      # a = "False"
b = str(True)       # b = "True"
c = str(5)         # c = "5"
d = str(5.7)       # d = "5.7"

Функция __str()__ может быть актуальна, например, при добавлении к строке значения другого типа. Например, в следующем случае мы получим ошибку:

In [64]:
age = 22
message = "Age: " + age     # Ошибка
print(message)

TypeError: can only concatenate str (not "int") to str

Если число складывается с число, то это стандартная операция сложения чисел. Если строка складывается со строкой, то это операция объединения строк. Но каким образом выполнить операцию сложения по отношение к строке и числу, Python не знает. И если мы в данном случае мы хотим выполнить операцию объединения строк, то число можно привести к строке с помощью функции str():

In [65]:
age = 22
message = "Age: " + str(age)   # Age: 22
print(message)

Age: 22


# Область видимости переменных
Область видимости или scope определяет контекст переменной, в рамках которого ее можно использовать. В Python есть два типа контекста: глобальный и локальный.
Глобальный контекст

Глобальный контекст подразумевает, что переменная является глобальной, она определена вне любой из функций и доступна любой функции в программе. Например:

In [66]:
name = "Tom"
 
def say_hi():
    print("Hello", name)
 
 
def say_bye():
    print("Good bye", name)

say_hi()
say_bye()

Hello Tom
Good bye Tom


Здесь переменная name является глобальной и имеет глобальную область видимости. И обе определенные здесь функции могут свободно ее использовать.
Локальный контекст

В отличие от глобальных переменных локальная переменная определяется внутри функции и доступна только из этой функции, то есть имеет локальную область видимости:

In [67]:
def say_hi():
    name = "Sam"
    surname = "Johnson"
    print("Hello", name, surname)
 
 
def say_bye():
    name = "Tom"
    print("Good bye", name)

say_hi()
say_bye()

Hello Sam Johnson
Good bye Tom


В данном случае в каждой из двух функций определяется локальная переменная name. И хотя эти переменные называются одинаково, но тем не менее это две разных переменных, каждая из которых доступна только в рамках своей функции. Также в функции say_hi() определена переменная surname, которая также является локальной, поэтому в функции say_bye() мы ее использовать не сможем.

### Скрытие переменных

Есть еще один вариант определения переменной, когда локальная переменная скрывают глобальную с тем же именем:

In [68]:
name = "Tom"
 
def say_hi():
    name = "Bob"        # скрываем значение глобальной переменной
    print("Hello", name)
 
 
def say_bye():
    print("Good bye", name)
 
 
say_hi()    # Hello Bob
say_bye()   # Good bye Tom

Hello Bob
Good bye Tom


Здесь определена глобальная переменная name. Однако в функции say_hi определена локальная переменная с тем же именем name. И если функция say_bye использует глобальную переменную, то функция say_hi использует локальную переменную, которая скрывает глобальную.

Если же мы хотим изменить в локальной функции глобальную переменную, а не определить локальную, то необходимо использовать ключевое слово global:

In [69]:
name = "Tom"
 
def say_hi():
    global  name
    name = "Bob"        # изменяем значение глобальной переменной
    print("Hello", name)
 
 
def say_bye():
    print("Good bye", name)
 
 
say_hi()    # Hello Bob
say_bye()   # Good bye Bob
nonlocal

SyntaxError: invalid syntax (3239105855.py, line 15)

Выражение __nonlocal__ прикрепляет идентификатор к переменной из ближайщего окружающего контекста (за исключением глобального контекста). Обычно nonlocal применяется во вложенных функциях, когда надо прикрепить идентификатор за переменной или параметром окружающей внешней функции. Рассмотрим ситуацию, где это выражение может пригодиться:

In [70]:
def outer():  # внешняя функция
    n = 5
 
    def inner():    # вложенная функция
        print(n)
 
    inner()     # 5
    print(n)
 
 
outer()     # 5

5
5


Здесь вложенная локальная функция inner() выводит на консоль значение переменной n, которая определена во внешней функции outer(). Затем в функции outer() вызывается внутренняя функция inner().

При вызове функции outer() здесь мы ожидаемо увидим на консоли два раза число 5. Однако в данном случае вложенная функция inner() просто получает значение. Теперь возьмем другую ситуацию, когда вложенная функция присваивает значение переменной:

In [71]:
def outer():  # внешняя функция
    n = 5
 
    def inner():    # вложенная функция
        n = 25
        print(n)
 
    inner()     # 25
    print(n)
 
 
outer()     # 5 
# 25    - inner
# 5     - outer

25
5


При присвоении значения во вложенной функции: n = 25 будет создаваться новая переменная n, которая скроет переменную n из окружающей внешней функции outer. В итоге мы получим при выводе два разных числа. Чтобы во вложенной функции указать, что идентификатор во вложенной функции будет представлять переменную из окружающей функции, применяется выражение nonlocal:

In [72]:
def outer():  # внешняя функция
    n = 5
 
    def inner():    # вложенная функция
        nonlocal n  # указываем, что n - это переменная из окружающей функции
        n = 25
        print(n)
 
    inner()     # 25
    print(n)
 
 
outer()          # 25

25
25


# Замыкания

__Замыкание (closure)__ представляет функцию, которая запоминает свое лексическое окружение даже в том случае, когда она выполняется вне своей области видимости.

Технически замыкание включает три компонента:

1) внешняя функция, которая определяет некоторую область видимости и в которой определены некоторые переменные и параметры - лексическое окружение

2) переменные и параметры (лексическое окружение), которые определены во внешней функции

3) вложенная функция, которая использует переменные и параметры внешней функции

Для определения замыканий в Python применяются локальные функции:

In [ ]:
def outer():        # внешняя функция
    n = 5           # лексическое окружение - локальная переменная
 
    def inner():      # локальная функция
        nonlocal n
        n += 1        # операции с лексическим окружением
        print(n)
 
    return inner
 
fn = outer()   # fn = inner, так как функция outer возвращает функцию inner
# вызываем внутреннюю функцию inner
fn()    # 6
fn()    # 7
fn()    # 8

Здесь функция outer определяет локальную переменную n - это и есть лексическое окружение для внутренней функции:

Внутри функции outer определена внутренняя функция - локальная функция inner, которая обращается к своему лексическому окружению - переменной n - увеличивает ее значение на единицу и выводит на консоль:

In [74]:
def inner():      # локальная функция
    nonlocal n
    n += 1        # операции с лексическим окружением
    print(n)

SyntaxError: no binding for nonlocal 'n' found (3398544855.py, line 2)

Эта локальная функция возвращается функцией outer:

```
return inner
```

В программе вызываем функцию outer и получаем в переменную fn локальную функцию inner:

In [76]:
fn = outer()

Переменная fn и представляет собой замыкание, то есть объединяет две вещи: функцию и окружение, в котором функция была создана. И несмотря на то, что мы получили локальную функцию и можем ее вызывать вне ее окружающей функции, в которой она определена, тем не менее она запомнила свое лексическое окружение и может к нему обращаться и изменять, что мы увидим по консольному выводу:

In [77]:
fn()    # 6
fn()    # 7
fn()    # 8

6
7
8


### Применение параметров

Кроме внешних переменных к лексическому окружению также относятся параметры окружающей функции. Рассмотрим использование параметров:

In [78]:
def multiply(n):
    def inner(m): return n * m
 
    return inner
 
fn = multiply(5)
print(fn(5))        # 25
print(fn(6))        # 30
print(fn(7))        # 35

25
30
35


Здесь внешняя функция - multiply возвращает функцию, которая принимает число и возвращает число.

Вызов функции multiply() возвращает локальную функцию inner:

In [79]:
def inner(m): return n * m

Эта функция запоминает окружение, в котором она была создана, в частности, значение параметра n. Кроме того, сама принимает параметр и возвращает произведение параметров n и m.

В итоге при вызове функции multiply определяется переменная fn, которая получает локальную функцию inner и ее лексическое окружение - значение параметра n:

In [80]:
fn = multiply(5)

В данном случае параметр n равен 5.

При вызове локальной функции, например, в случае:

In [81]:
print(fn(6))        # 30

30


Число 6 передается для параметра m локальной функции, которая возвращает произведение n и m, то есть 5 * 6 = 30.

Также можно было бы сократить этот код с помощью лямбд:

In [82]:
def multiply(n): return lambda m: n * m
 
fn = multiply(5)
print(fn(5))        # 25
print(fn(6))        # 30
print(fn(7))        # 35

25
30
35


# Декораторы
__Декораторы__ в Python представляют функцию, которая в качестве параметра получает функцию и в качестве результата также возвращает функцию. Декораторы позволяют модифицировать выполняемую функцию, значения ее параметров и ее результат без изменения исходного кода этой функции.

Рассмотрим простейший пример:

In [83]:
# определение функции декоратора
def select(input_func):    
    def output_func():      # определяем функцию, которая будет выполняться вместо оригинальной
        print("*****************")  # перед выводом оригинальной функции выводим всякую звездочки
        input_func()                # вызов оригинальной функции
        print("*****************")  # после вывода оригинальной функции выводим всякую звездочки
    return output_func     # возвращаем новую функцию
 
# определение оригинальной функции
@select         # применение декоратора select
def hello():
    print("Hello METANIT.COM")
# вызов оригинальной функции
hello()

*****************
Hello METANIT.COM
*****************


Вначале определяется собственно функция декоратора, которая в данном случае называется select(). В качестве параметра декоратор получает функцию (в данном случае параметр input_func), к которой этот декоратор будет применяться:

In [84]:
def select(input_func):    
    def output_func():      # определяем функцию, которая будет выполняться вместо оригинальной
        print("*****************")  # перед выводом оригинальной функции выводим всякую звездочки
        input_func()                # вызов оригинальной функции
        print("*****************")  # после вывода оригинальной функции выводим всякую звездочки
    return output_func     # возвращаем новую функцию

Результатом декоратора в данном случае является локальная функция output_func, в которой вызывается входная функция input_func. Для простоты здесь перед и после вызыва input_func для красоты просто выводим набор символов "#".

Далее определяется стандартная функция, к которой применяется декоратор - в данном случае это функция hello, которая просто выводит на консоль некоторую строку:

In [85]:
@select         # применение декоратора select
def hello():
    print("Hello METANIT.COM")

Для применения декоратора перед определением функции указывается символ @, после которого идет имя декоратора. То есть в данном случае к функции hello() применяется декоратор select().

Далее вызываем обычную функцию:

In [86]:
hello()

*****************
Hello METANIT.COM
*****************


Поскольку к этой функции применяется декоратор select, то в результате функциия hello передается в декоратор select() в качестве параметра input_func. И поскольку декоратор возвращает новую функцию - output_func, то фактически в данном случае будет выполняться именно эта функция output_func()

В итоге мы получим следующий консольный вывод:
```
*****************
Hello METANIT.COM
*****************
```
### Получение параметров функции в декораторе

Декоратор может перехватывать передаваемые в функцию аргументы:

In [87]:
# определение функции декоратора
def check(input_func):    
    def output_func(*args):      # через *args получаем значения параметров оригинальной функции
        input_func(*args)                # вызов оригинальной функции
    return output_func     # возвращаем новую функцию
 
# определение оригинальной функции
@check
def print_person(name, age):
    print(f"Name: {name}  Age: {age}")

# вызов оригинальной функции
print_person("Tom", 38)

Name: Tom  Age: 38


Здесь функция print_person() принимает два параметра: name (имя) и age (возраст). К этой функции применяется декоратор check()

В декораторе check возвращается локальная функция output_func(), которая принимает некоторый набор значений в виде параметра `*args` - это те значения, которые передаются в оригинальную функцию, к которой применяется декоратор. То есть в данном случае `*args` будет содержать значения параметров name и age.

In [ ]:
def check(input_func):    
    def output_func(*args):      # через *args получаем значения параметров функции input_func

Здесь просто передаем эти значения в оригинальную функцию:

In [ ]:
input_func(*args)

В итоге в данном получим следующий консольный вывод
```
Name: Tom  Age: 38
```
Но что, если в функцию print_person будет передано какое-то недопустимое значение, например, отрицательный возраст? Одним из преимуществ декораторов как раз является то, что мы можем проверить и при необходимости модифицировать значения параметров. Например:

In [90]:
# определение функции декоратора
def check(input_func):    
    def output_func(*args):
        name = args[0]
        age = args[1]           # получаем значение второго параметра
        if age < 0: age = 1     # если возраст отрицательный, изменяем его значение на 1
        input_func(name, age)   # передаем функции значения для параметров
    return output_func
 
# определение оригинальной функции
@check
def print_person(name, age):
    print(f"Name: {name}  Age: {age}")

# вызов оригинальной функции
print_person("Tom", 38)
print_person("Bob", -5)

Name: Tom  Age: 38
Name: Bob  Age: 1


args фактически представляет набор значений, и, используя индексы, мы можем получить значения параметров по позиции и что-то с ними сделать. Так, здесь, если значение возраста меньше 0, то устанавливаем 1. Затем передаем эти значения в вызов функции. В итоге здесь получим следующий вывод:
```
Name: Tom  Age: 38
Name: Bob  Age: 1
```
### Получение результата функции

Подобным образом можно получить результат функции и при необходимости изменить его:

In [91]:
# определение функции декоратора
def check(input_func):    
    def output_func(*args):
        result = input_func(*args)   # передаем функции значения для параметров
        if result < 0: result = 0   # если результат функции меньше нуля, то возвращаем 0
        return result
    return output_func
 
# определение оригинальной функции
@check
def sum(a, b):
    return a + b
 
# вызов оригинальной функции
result1 = sum(10, 20)
print(result1)          # 30
 
result2 = sum(10, -20)
print(result2)          # 0

30
0


Здесь определена функция sum(), которая возвращает сумму чисел. В декораторе check проверяем результат функции и для простоты, если он меньше нуля, то возвращаем 0.